From cac9aa2b469f5417bb1433b3461281b8f3248c29 Mon Sep 17 00:00:00 2001
From: Deva S <devadharuns@google.com>
Date: Wed, 30 Sep 2026 12:59:07 +0000
Subject: [PATCH] libcxxabi: declare __gnu_unwind_frame in cxa_personality
 (#189787)

ARM EHABI builds of libcxxabi fail with clang-22+ because
cxa_personality.cpp calls __gnu_unwind_frame without a visible
declaration, triggering:

  error: use of undeclared identifier '__gnu_unwind_frame'

Add an extern "C" forward declaration before the EHABI unwind helper so
the source compiles correctly.

BUG=b:562535026
TEST=cros_setup_toolchains

patch.cherry: true
patch.metadata.author: Khem Raj <khem.raj@oss.qualcomm.com>
patch.metadata.original_sha: 8119f1854948b50358bbfaea08f207f51970f06c
patch.platforms: chromiumos
patch.version_range.from: 600745
patch.version_range.until: 616126

---
 libcxxabi/src/cxa_personality.cpp | 3 +++
 1 file changed, 3 insertions(+)

diff --git a/libcxxabi/src/cxa_personality.cpp b/libcxxabi/src/cxa_personality.cpp
index 35f4e3807724..dcd01ab49f15 100644
--- a/libcxxabi/src/cxa_personality.cpp
+++ b/libcxxabi/src/cxa_personality.cpp
@@ -1116,6 +1116,9 @@ __gxx_personality_seh0(PEXCEPTION_RECORD ms_exc, void *this_frame,
 
 #else
 
+extern "C" _Unwind_Reason_Code __gnu_unwind_frame(_Unwind_Exception*,
+                                                  _Unwind_Context*);
+
 // Helper function to unwind one frame.
 // ARM EHABI 7.3 and 7.4: If the personality function returns _URC_CONTINUE_UNWIND, the
 // personality routine should update the virtual register set (VRS) according to the
-- 
2.39.5